hack/e2e.go

```diff
@@ -14,968 +14,138 @@ See the License for the specific language governing permissions and
-	bp := filepath.Join(location, "gcp-resources-before.txt")
-	up := filepath.Join(location, "gcp-resources-cluster-up.txt")
-	cdp := filepath.Join(location, "gcp-resources-cluster-down.txt")
-	ap := filepath.Join(location, "gcp-resources-after.txt")
-	dp := filepath.Join(location, "gcp-resources-diff.txt")
-
-	if err := ioutil.WriteFile(bp, before, mode); err != nil {
-		return err
-	}
-	if err := ioutil.WriteFile(up, clusterUp, mode); err != nil {
+func wait(cmd string, args ...string) error {
+	c := exec.Command(cmd, args...)
+	c.Stdout = os.Stdout
+	c.Stderr = os.Stderr
+	if err := c.Start(); err != nil {
 		return err
 	}
-	if err := ioutil.WriteFile(cdp, clusterDown, mode); err != nil {
-		return err
-	}
-	if err := ioutil.WriteFile(ap, after, mode); err != nil {
-		return err
-	}
-
-	cmd := exec.Command("diff", "-sw", "-U0", "-F^\\[.*\\]$", bp, ap)
-	if *verbose {
-		cmd.Stderr = os.Stderr
-	}
-	stdout, cerr := cmd.Output()
-	if err := ioutil.WriteFile(dp, stdout, mode); err != nil {
-		return err
-	}
-	if cerr == nil { // No diffs
-		return nil
-	}
-	lines := strings.Split(string(stdout), "\n")
-	if len(lines) < 3 { // Ignore the +++ and --- header lines
-		return nil
-	}
-	lines = lines[2:]
-
-	var added, report []string
-	resourceTypeRE := regexp.MustCompile(`^@@.+\s(\[\s\S+\s\])$`)
-	for _, l := range lines {
-		if matches := resourceTypeRE.FindStringSubmatch(l); matches != nil {
-			report = append(report, matches[1])
-		}
-		if strings.HasPrefix(l, "+") && len(strings.TrimPrefix(l, "+")) > 0 {
-			added = append(added, l)
-			report = append(report, l)
-		}
-	}
-	if len(added) > 0 {
-		return fmt.Errorf("Error: %d leaked resources\n%v", len(added), strings.Join(report, "\n"))
-	}
-	return nil
-}
-
-func ListResources() ([]byte, error) {
-	log.Printf("Listing resources...")
-	cmd := exec.Command("./cluster/gce/list-resources.sh")
-	if *verbose {
-		cmd.Stderr = os.Stderr
-	}
-	stdout, err := cmd.Output()
-	if err != nil {
-		return nil, fmt.Errorf("Failed to list resources (%s):\n%s", err, string(stdout))
-	}
-	return stdout, nil
-}
-
-func Build() error {
-	// The build-release script needs stdin to ask the user whether
-	// it's OK to download the docker image.
-	cmd := exec.Command("make", "quick-release")
-	cmd.Stdin = os.Stdin
-	if err := finishRunning(cmd); err != nil {
-		return fmt.Errorf("error building kubernetes: %v", err)
-	}
-	return nil
-}
-
-type deployer interface {
-	Up() error
-	IsUp() error
-	SetupKubecfg() error
-	Down() error
-}
-
-func getDeployer() (deployer, error) {
-	switch *deployment {
-	case "bash":
-		return bash{}, nil
-	case "kops":
-		return NewKops()
-	case "kubernetes-anywhere":
-		return NewKubernetesAnywhere()
-	default:
-		return nil, fmt.Errorf("Unknown deployment strategy %q", *deployment)
-	}
+	return c.Wait()
 }
 
-type bash struct{}
-
-func (b bash) Up() error {
-	return finishRunning(exec.Command("./hack/e2e-internal/e2e-up.sh"))
+// Struct that allows unit tests to override functionality.
```

Comment:
I usually see this done with a global variable for small packages. Something like `var osStat = os.Stat` and then in tests override it.